### ISSUE TITLE
Execute notification functions in their own threads

### ISSUE THREAD

**Issue by pdnelson** [YCPRadioTelescope/YCP-RT-ControlRoom#209]
Presently, push notifications and emails are being executed in series with all of the code in which they are implemented. This can be problematic when quick execution is necessary, and since both emails and push notifications retrieve lists from the database, and execute on each retrieved element, this can slow things down.
For example, when a limit switch is hit, we want the recovery of that to be instantaneous, but it will be held up if the program has to send out emails and push notifications. This is especially important in the `ProductionPLCDriver`, though we will want to use it everywhere.

The solution to this is to execute each notification function in its own lower-priority thread, so that it runs in parallel with important code, as opposed to in series. Each thread should be terminated at the end of its run to avoid memory leaks. Push notifications and Emails should have their own threads. Each function that the thread calls can be located in its respective file.

Things that will need looked into:
- Passing parameters to threads
- Abstraction of threads so the same function can be used anywhere
- Thread priority. Notification threads should be a lower priority than other business logic

**Comment by pdnelson** [comment]
While researching into this, I found that it would be a better practice to go about this using tasks as opposed to threads. Through the use of tasks, this allowed the code to be testable as it was before the task changes because tasks can return results, which is much more difficult to achieve with threads. We can, technically, return results from a block of code executed with a thread, but doing so would be needlessly difficult, when C# has a built-in way to handle these situations.

**Comment by pdnelson** [comment]
Apparently tasks aren't guaranteed to run in parallel in the code, so a thread was needed anyway. This was fine, though, and allowed us to use `ThreadPriority`, which I think is a good idea.
